doc/help/kdb-record.md

```diff
@@ -0,0 +1,144 @@
+
+- `ELEKTRA_RECORD_ACTIVE_SESSION`: If this environment variable is set, the specified session will be the active session.
+
+## COMMON PARAMETERS
+
+Most commands allow specification of the name of the session they should work against.
+If no session name is provided, the active session will be used.
+
+A notable exception is the `start` command.
+If no session name is provided, it will create a random name.
+
+## COMMANDS
+
+### `start [session_name]`
+The `start` command is used to start a new or resume an existing session.
+If the environment variable `ELEKTRA_RECORD_ACTIVE_SESSION` is set, this command has no immediate effect.
+As soon as the environment variable is unset, the session started or resumed with this command will be active.
+
+This command will accept a single optional argument that is the name of the session.
+If the name is not provided, a random name will be generated.
+
+On success, the name of the session will be printed to `stdout`, and the return code is `0`.
+
+### `stop`
+The `stop` command is used to stop the active recording session.
+The session can then be resumed using the `start` command with the appropriate session name.
+Stopping a recording session will not delete any recorded changes.
+
+### `reset [session_name]`
+The `reset` command will clear the specified recording.
+If no session name argument is provided, the currently active session will be reset.
+
+### `delete [session_name]`
+Deletes the specified recording session.
+It the optional argument is not provided, the current session will be deleted.
+If it is the currently active session, it will be stopped.
+
+### `mark [--meta meta_name] <key_name>`
+This command marks a key to be used for validation when importing the recording.
+By default, the values of a marked key will be compared when importing.
+The import will fail if the key has a different value.
+
+Using the `--meta` flag, one can specify to use the value of a specific meta instead of the value of the key.
+
+### `list`
+Lists all known recording sessions.
+
+### `name`
+Prints out the name of the currently active recording session.
+If no session is active, the return code is `1`, otherwise `0`.
+
+### `changes [session_name]`
+Shows all the changes in the specified session in human-readable format.
+Outputs to `stdout`.
+If no session is specified, the currently active session is used.
+
+### `export  [--replay|--diff] [session_name]`
+Exports the changes made in the specified session.
+Outputs to `stdout`.
+If no session is specified, the currently active session is used.
+
+The `export` command has the ability to output two different types: diffs and replays.
+
+The `--diff` flag (default) only exports what has actually changed in the recording session.
+This means, e.g., that when you add a key and remove it in the same session, it will not be exported.
+
+The `--replay` flag creates a file that contains every modifying operation made in the session in the order they were performed.
```

Comment:
> Why would this be needed?

From a purely technical standpoint: There is no way of knowing what plugins do during `kdbSet`. Therefore, we cannot guarantee that a full replay of the individual `kdbSet` changes has the same result (especially in regard to  side effects) as a single collapsed diff.

Take for example this sequence of commands:

```
kdb set user:/foo 123
kdb record start
kdb rm user:/foo
kdb set user:/foo 123
kdb record stop
```

The full replay would remove and then newly create `user:/foo`. The collapsed diff on the other hand would be empty. Especially such a case could have big differences in side effects. Imagine, if the config is backed by an SQL database and there are triggers attached to the tables that store the config. The full replay and the collapsed diff would cause different triggers to run.

Now from a philosophical POV, you could probably argue that both variants should have the same effect. But I don't think we can assume that.

There is also a second technical reason: Recording the replay log and collapsing it into a diff is more efficient in terms of storage. It is similar to a Copy-on-Write filesystem. We only record what actually changed. And if we already record the log, we might as well provide an option to export and replay it. The other option would be, not recording the log and instead producing a diff based on saved snapshots. However, with the current setup (`kdb record` records changes to the whole KDB), every `kdb start` would have to make a complete copy of the whole KDB, so that we can later produce a diff.

> Isn't there an inconsistency with --format=replay vs. --replay?

I think that's a mistake, from a recent commit (possibly you also look at different versions of the file).

Reply:
Re https://github.com/ElektraInitiative/libelektra/issues/4401#issuecomment-1179551907


> We would only store keys that changed/are relevant. So every change makes at most two keys: in base and in ours (or in assert).

I clearly misunderstood what you meant by base+ours+assert. I thought you want to collect base only once at the start, but since we don't know what will change, we need to collect everything. I explained this in my original response above ("The other option would be [...]").

Turns out your solution is actually very similar to the "collect a log of everything" solution. The only difference seems to be whether we remember intermediate values during an active session, or just the initial and final values. But why would that mean we need two separate implementations (suggested in https://github.com/ElektraInitiative/libelektra/issues/4401#issuecomment-1179526126)?